x/bank/keeper/send.go

```diff
@@ -22,7 +22,11 @@ import (
 type SendKeeper interface {
 	ViewKeeper
 
-	InputOutputCoins(ctx context.Context, inputs types.Input, outputs []types.Output) error
+	AppendSendRestriction(restriction types.SendRestrictionFn)
```

Comment:
Are send restrictions meant to be updated as the chain is running? Because I think having them initted once reduces API surface.

Reply:
They're not meant to be updated as the chain is running. The reasons they're appended/prepended like this are:
1. The bank keeper can be created as one of the first (crucial since several other modules depend on it). It's expected that modules that need to inject send restrictions will often also need other bank keeper functionality.
2. It allows other modules to inject their own restrictions (e.g. in their own keeper constructors) without needing knowledge of any other modules that have injected/will inject restrictions.
3. Modules that need to inject a restriction won't work without that restriction being applied. It should be up to that module to make sure the restriction is injected. It shouldn't be relegated to something a blockchain author needs to know to wire up outside the module. If the restriction can be initialized only once, you'd have to create all the keepers, know which ones have restriction to apply, combine them, and give them and init them.

For example, we (Provenance) have three modules with send restrictions: [sanction](https://github.com/provenance-io/cosmos-sdk/blob/main-pio/x/sanction/keeper/keeper.go#L52), [quarantine](https://github.com/provenance-io/cosmos-sdk/blob/main-pio/x/quarantine/keeper/keeper.go#L34), and [marker](https://github.com/provenance-io/provenance/blob/main/x/marker/keeper/keeper.go#L127). None of those modules work correctly without those send restrictions being applied. Because they're injected into the bank module during their own keeper creation, we can be certain that they're being applied.

Also, because they're being injected during their own keeper creation, they can each have unit tests in their own module that make sure they're being applied. If they had to be initialized once all together, those tests would have to live live at an app level.

Basically, this design fixes some pain points I encountered by having multiple modules with gov hooks.